File: platform/packages_py/fetch_httpx/logger.py
"""
Logger Module - Defensive Programming Logging Interface

This module provides a standardized logging interface for the fetch_httpx package.
It supports structured logging with JSON and human-readable formats, log level filtering,
and can be overwritten via constructor arguments.

Usage:
    from . import logger as logger_module
    logger = logger_module.create("fetch_httpx", __file__)

    logger.info("Request started", context={"url": url, "method": method})
    logger.debug("Headers resolved", context={"headers": headers})
    logger.error("Request failed", context={"error": str(e)})

Environment Variables:
    LOG_LEVEL: TRACE, DEBUG, INFO, WARN, ERROR, SILENT (default: INFO)
    LOG_FORMAT: json, pretty (default: pretty in dev, json in prod)
    PYTHON_ENV: development, production (default: development)
"""

from __future__ import annotations

import json
import os
import sys
from collections.abc import Callable
from datetime import UTC, datetime
from enum import IntEnum
from pathlib import Path
from typing import Any


class LogLevel(IntEnum):
    """Log level enumeration with numeric values for comparison."""
    TRACE = 0
    DEBUG = 10
    INFO = 20
    WARN = 30
    ERROR = 40
    SILENT = 100


# Type aliases
LogFormatter = Callable[["LogEntry"], str]
LogOutput = Callable[[str], None]


class LogEntry:
    """Structured log entry containing all log metadata."""

    __slots__ = ("timestamp", "level", "package", "file", "message", "context", "correlation_id")

    def __init__(
        self,
        timestamp: str,
        level: str,
        package: str,
        file: str,
        message: str,
        context: dict[str, Any] | None = None,
        correlation_id: str | None = None,
    ) -> None:
        self.timestamp = timestamp
        self.level = level
        self.package = package
        self.file = file
        self.message = message
        self.context = context or {}
        self.correlation_id = correlation_id

    def to_dict(self) -> dict[str, Any]:
        """Convert log entry to dictionary representation."""
        result = {
            "timestamp": self.timestamp,
            "level": self.level,
            "package": self.package,
            "file": self.file,
            "message": self.message,
        }
        if self.context:
            result["context"] = self.context
        if self.correlation_id:
            result["correlation_id"] = self.correlation_id
        return result


class Logger:
    """
    Standard logging interface for fetch_httpx package.

    Provides defensive programming patterns with structured logging,
    configurable output format, and log level filtering.

    Args:
        package_name: Name of the package (e.g., "fetch_httpx")
        filename: Source filename (typically __file__)
        level: Minimum log level to output
        formatter: Custom formatter function (LogEntry -> str)
        output: Custom output function (str -> None)
        correlation_id: Optional correlation ID for request tracing
    """

    def __init__(
        self,
        package_name: str,
        filename: str,
        level: LogLevel = LogLevel.INFO,
        formatter: LogFormatter | None = None,
        output: LogOutput | None = None,
        correlation_id: str | None = None,
    ) -> None:
        self.package_name = package_name
        self.filename = self._extract_filename(filename)
        self.level = level
        self.formatter = formatter or self._get_default_formatter()
        self.output = output or self._default_output
        self.correlation_id = correlation_id

    @staticmethod
    def _extract_filename(filepath: str) -> str:
        """Extract just the filename from a full path."""
        return Path(filepath).name

    @staticmethod
    def _get_default_formatter() -> LogFormatter:
        """Get default formatter based on environment."""
        env = os.environ.get("PYTHON_ENV", "development").lower()
        log_format = os.environ.get("LOG_FORMAT", "").lower()

        if log_format == "json" or (env == "production" and log_format != "pretty"):
            return Logger._json_formatter
        return Logger._pretty_formatter

    @staticmethod
    def _json_formatter(entry: LogEntry) -> str:
        """Format log entry as JSON (production)."""
        return json.dumps(entry.to_dict(), separators=(",", ":"))

    @staticmethod
    def _pretty_formatter(entry: LogEntry) -> str:
        """Format log entry as human-readable (development)."""
        # Color codes for different log levels
        colors = {
            "TRACE": "\033[90m",   # Gray
            "DEBUG": "\033[36m",   # Cyan
            "INFO": "\033[32m",    # Green
            "WARN": "\033[33m",    # Yellow
            "ERROR": "\033[31m",   # Red
        }
        reset = "\033[0m"
        color = colors.get(entry.level, "")

        # Format: [TIMESTAMP] LEVEL [package:file] message {context}
        parts = [
            f"[{entry.timestamp}]",
            f"{color}{entry.level:5}{reset}",
            f"[{entry.package}:{entry.file}]",
            entry.message,
        ]

        if entry.correlation_id:
            parts.append(f"(cid:{entry.correlation_id})")

        if entry.context:
            ctx_str = " ".join(f"{k}={_format_value(v)}" for k, v in entry.context.items())
            parts.append(f"{{ {ctx_str} }}")

        return " ".join(parts)

    @staticmethod
    def _default_output(formatted: str) -> None:
        """Default output to stderr."""
        print(formatted, file=sys.stderr)

    def _create_entry(
        self,
        level: str,
        message: str,
        context: dict[str, Any] | None = None,
    ) -> LogEntry:
        """Create a log entry with current timestamp."""
        return LogEntry(
            timestamp=datetime.now(UTC).isoformat(timespec="milliseconds"),
            level=level,
            package=self.package_name,
            file=self.filename,
            message=message,
            context=context,
            correlation_id=self.correlation_id,
        )

    def _log(
        self,
        level: LogLevel,
        message: str,
        context: dict[str, Any] | None = None,
    ) -> None:
        """Internal log method with level checking."""
        if level < self.level:
            return

        entry = self._create_entry(level.name, message, context)
        formatted = self.formatter(entry)
        self.output(formatted)

    def trace(self, message: str, context: dict[str, Any] | None = None) -> None:
        """Log at TRACE level (most verbose)."""
        self._log(LogLevel.TRACE, message, context)

    def debug(self, message: str, context: dict[str, Any] | None = None) -> None:
        """Log at DEBUG level."""
        self._log(LogLevel.DEBUG, message, context)

    def info(self, message: str, context: dict[str, Any] | None = None) -> None:
        """Log at INFO level."""
        self._log(LogLevel.INFO, message, context)

    def warn(self, message: str, context: dict[str, Any] | None = None) -> None:
        """Log at WARN level."""
        self._log(LogLevel.WARN, message, context)

    def error(self, message: str, context: dict[str, Any] | None = None) -> None:
        """Log at ERROR level."""
        self._log(LogLevel.ERROR, message, context)

    def with_correlation_id(self, correlation_id: str) -> Logger:
        """Create a new logger instance with the specified correlation ID."""
        return Logger(
            package_name=self.package_name,
            filename=self.filename,
            level=self.level,
            formatter=self.formatter,
            output=self.output,
            correlation_id=correlation_id,
        )

    def child(self, filename: str) -> Logger:
        """Create a child logger for a different file in the same package."""
        return Logger(
            package_name=self.package_name,
            filename=filename,
            level=self.level,
            formatter=self.formatter,
            output=self.output,
            correlation_id=self.correlation_id,
        )


def _format_value(value: Any) -> str:
    """Format a value for pretty printing, handling special cases."""
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, str):
        # Truncate long strings
        if len(value) > 100:
            return f'"{value[:97]}..."'
        return f'"{value}"'
    if isinstance(value, dict | list):
        s = json.dumps(value, separators=(",", ":"))
        if len(s) > 100:
            return f"{s[:97]}..."
        return s
    return str(value)


def _get_level_from_env() -> LogLevel:
    """Get log level from environment variable."""
    level_str = os.environ.get("LOG_LEVEL", "INFO").upper()
    try:
        return LogLevel[level_str]
    except KeyError:
        return LogLevel.INFO


def create(
    package_name: str,
    filename: str,
    level: LogLevel | None = None,
    formatter: LogFormatter | None = None,
    output: LogOutput | None = None,
) -> Logger:
    """
    Factory function to create a logger instance.

    Args:
        package_name: Name of the package (e.g., "fetch_httpx")
        filename: Source filename (typically __file__)
        level: Override log level (default: from LOG_LEVEL env var)
        formatter: Custom formatter function
        output: Custom output function

    Returns:
        Configured Logger instance

    Example:
        logger = create("fetch_httpx", __file__)
        logger.info("Client created", context={"base_url": base_url})
    """
    resolved_level = level if level is not None else _get_level_from_env()
    return Logger(
        package_name=package_name,
        filename=filename,
        level=resolved_level,
        formatter=formatter,
        output=output,
    )


# Convenience: Create a default logger for this module
_module_logger = create("fetch_httpx", __file__)


def get_module_logger() -> Logger:
    """Get the default module logger."""
    return _module_logger


__all__ = [
    "LogLevel",
    "LogEntry",
    "Logger",
    "create",
    "get_module_logger",
]

File: platform/packages_mjs/fetch_undici/src/logger.ts
/**
 * Logger Infrastructure for fetch-undici
 *
 * Provides structured logging with verbose output for defensive programming.
 * Usage: const log = logger.create('fetch-undici', import.meta.url)
 */

/** Log levels in order of verbosity */
export type LogLevel = 'trace' | 'debug' | 'info' | 'warn' | 'error' | 'silent'

/** Log level numeric values for comparison */
const LOG_LEVELS: Record<LogLevel, number> = {
  trace: 0,
  debug: 1,
  info: 2,
  warn: 3,
  error: 4,
  silent: 5
}

/** Structure of a log entry */
export interface LogEntry {
  level: LogLevel
  timestamp: string
  package: string
  file: string
  message: string
  context?: Record<string, unknown>
}

/** Options for logger creation */
export interface LoggerOptions {
  level?: LogLevel
  transport?: (entry: LogEntry) => void
  json?: boolean
  redact?: string[]
}

/** Logger interface with level methods */
export interface Logger {
  trace(message: string, context?: Record<string, unknown>): void
  debug(message: string, context?: Record<string, unknown>): void
  info(message: string, context?: Record<string, unknown>): void
  warn(message: string, context?: Record<string, unknown>): void
  error(message: string, context?: Record<string, unknown>): void
  child(context: Record<string, unknown>): Logger
  setLevel(level: LogLevel): void
  getLevel(): LogLevel
}

/** Default keys to redact from logs */
const DEFAULT_REDACT_KEYS = [
  'password',
  'secret',
  'token',
  'apiKey',
  'api_key',
  'authorization',
  'auth',
  'credential',
  'private'
]

/**
 * Redact sensitive values from context object
 */
function redactSensitive(
  obj: Record<string, unknown>,
  redactKeys: string[]
): Record<string, unknown> {
  const result: Record<string, unknown> = {}

  for (const [key, value] of Object.entries(obj)) {
    const lowerKey = key.toLowerCase()
    const shouldRedact = redactKeys.some((k) => lowerKey.includes(k.toLowerCase()))

    if (shouldRedact) {
      result[key] = '[REDACTED]'
    } else if (value && typeof value === 'object' && !Array.isArray(value)) {
      result[key] = redactSensitive(value as Record<string, unknown>, redactKeys)
    } else {
      result[key] = value
    }
  }

  return result
}

/**
 * Extract filename from import.meta.url or string path
 */
function extractFilename(filename: string | URL): string {
  if (filename instanceof URL) {
    const pathname = filename.pathname
    return pathname.split('/').pop() || 'unknown'
  }

  if (filename.startsWith('file://')) {
    try {
      const url = new URL(filename)
      return url.pathname.split('/').pop() || 'unknown'
    } catch {
      return filename.split('/').pop() || 'unknown'
    }
  }

  return filename.split('/').pop() || 'unknown'
}

/**
 * Get log level from environment variable
 */
function getEnvLogLevel(): LogLevel {
  const envLevel = process.env['LOG_LEVEL']?.toLowerCase()
  if (envLevel && envLevel in LOG_LEVELS) {
    return envLevel as LogLevel
  }
  return 'info'
}

/**
 * Default console transport
 */
function defaultTransport(entry: LogEntry, json: boolean): void {
  const output = json ? JSON.stringify(entry) : formatLogEntry(entry)

  switch (entry.level) {
    case 'trace':
    case 'debug':
      // eslint-disable-next-line no-console
      console.debug(output)
      break
    case 'info':
      // eslint-disable-next-line no-console
      console.info(output)
      break
    case 'warn':
      // eslint-disable-next-line no-console
      console.warn(output)
      break
    case 'error':
      // eslint-disable-next-line no-console
      console.error(output)
      break
  }
}

/**
 * Format log entry for human-readable output
 */
function formatLogEntry(entry: LogEntry): string {
  const { timestamp, level, package: pkg, file, message, context } = entry
  const levelStr = level.toUpperCase().padEnd(5)
  const prefix = `${timestamp} [${levelStr}] ${pkg}/${file}:`

  if (context && Object.keys(context).length > 0) {
    return `${prefix} ${message} ${JSON.stringify(context)}`
  }

  return `${prefix} ${message}`
}

/**
 * Create a logger instance
 */
function createLoggerInstance(
  packageName: string,
  filename: string | URL,
  options: LoggerOptions = {},
  parentContext: Record<string, unknown> = {}
): Logger {
  const file = extractFilename(filename)
  let currentLevel = options.level ?? getEnvLogLevel()
  const transport = options.transport
  const json = options.json ?? (process.env['LOG_FORMAT'] === 'json')
  const redactKeys = [...DEFAULT_REDACT_KEYS, ...(options.redact ?? [])]

  function shouldLog(level: LogLevel): boolean {
    return LOG_LEVELS[level] >= LOG_LEVELS[currentLevel]
  }

  function log(level: LogLevel, message: string, context?: Record<string, unknown>): void {
    if (!shouldLog(level)) return

    const mergedContext = { ...parentContext, ...context }
    const redactedContext =
      Object.keys(mergedContext).length > 0 ? redactSensitive(mergedContext, redactKeys) : undefined

    const entry: LogEntry = {
      level,
      timestamp: new Date().toISOString(),
      package: packageName,
      file,
      message,
      context: redactedContext
    }

    if (transport) {
      transport(entry)
    } else {
      defaultTransport(entry, json)
    }
  }

  return {
    trace: (message, context) => log('trace', message, context),
    debug: (message, context) => log('debug', message, context),
    info: (message, context) => log('info', message, context),
    warn: (message, context) => log('warn', message, context),
    error: (message, context) => log('error', message, context),

    child(context: Record<string, unknown>): Logger {
      return createLoggerInstance(packageName, filename, options, {
        ...parentContext,
        ...context
      })
    },

    setLevel(level: LogLevel): void {
      currentLevel = level
    },

    getLevel(): LogLevel {
      return currentLevel
    }
  }
}

/**
 * Logger factory
 *
 * @param packageName - Name of the package (e.g., 'fetch-undici')
 * @param filename - Module filename, typically import.meta.url
 * @param options - Optional logger configuration
 * @returns Logger instance
 *
 * @example
 * ```typescript
 * import { logger } from './logger.js'
 * const log = logger.create('fetch-undici', import.meta.url)
 *
 * log.debug('Request starting', { method: 'GET', url: '/users' })
 * log.info('Client created', { baseUrl: 'https://api.example.com' })
 * log.error('Request failed', { error: err.message })
 * ```
 */
export function create(
  packageName: string,
  filename: string | URL,
  options?: LoggerOptions
): Logger {
  return createLoggerInstance(packageName, filename, options)
}

/** Logger factory object */
export const logger = {
  create,
  LOG_LEVELS,
  getEnvLogLevel
}

export default logger
